project/TaglessGen.scala

```diff
@@ -368,54 +378,32 @@ class TaglessGen(
       |
       |  // to support shifting blocking operations to another pool.
       |  val contextShiftM: ContextShift[M]
-      |  val blocker: Blocker
       |
       | ${managed.map(interpreterDef).mkString("\n  ")}
       |  // Some methods are common to all interpreters and can be overridden to change behavior globally.
       | 
-      |  def primitive[J, A](f: J => A): Kleisli[M, J, A] = Kleisli { a =>
-      |   blocker.blockOn[M, A](try {
-      |    asyncM.delay(f(a))
-      |     } catch {
-      |       case scala.util.control.NonFatal(e) => asyncM.raiseError(e)
-      |     })(contextShiftM)
-      |   }
-      |  def primitive1[J, A](f: => A): M[A] =
-      |   blocker.blockOn[M, A](try {
-      |     asyncM.delay(f)
-      |   } catch {
-      |     case scala.util.control.NonFatal(e) => asyncM.raiseError(e)
-      |   })(contextShiftM)
+      |  def primitive[J, A](f: J => A): Kleisli[M, J, A] = Kleisli(a => primitive1(f(a)))
```

Comment:
if `f(a)` invoked outside the effect ? if yes, we might change this

Reply:
f(a) is invoked inside `primitive1` function
`primitive1` function is calling `f(a)` inside `delay` which already catches the exceptions:

All the code for catch the NonFatal is unneccessary if `f(a)` is suspended inside the `delay` like in this case.
A proof:
```
import cats.data._

class Interpreter[M[_]: Async] {
  val asyncM = Async[M]
  def primitive[J, A](f: J => A): Kleisli[M, J, A] = Kleisli(a => primitive1(f(a)))
  def primitive1[J, A](f: =>A): M[A] = asyncM.delay(f)
}

trait Client
trait Response
val ioInterpreter = new Interpreter[IO]

val clientResponseK: Kleisli[IO, Client, Response] = ioInterpreter.primitive[Client, Response](_ => throw new RuntimeException)
// clientResponseK: Kleisli[IO, Client, Response] = Kleisli(<function1>)
clientResponseK(new Client{}).unsafeRunSync
// thows java.lang.RuntimeException

val response: IO[Response] = ioInterpreter.primitive1(throw new RuntimeException)
// response: IO[Response] = Delay(<function0>)
response.unsafeRunSync
// thows java.lang.RuntimeException
```

As you see the Exception is suspended into the effect and not raised until you run it

Reply:
Seems monix.bio has same behaviour:
https://github.com/monix/monix-bio/blob/master/core/shared/src/main/scala/monix/bio/IO.scala#L3007-L3018

Reply:
right,
i'm happy to accept the PR, but i have some troubles on CI. seems like test is hanging

Reply:
https://github.com/laserdisc-io/fs2-aws/tree/series/3.x.x

Reply:
@semenodm I have checked and the problem is coming from this previous commit https://github.com/laserdisc-io/fs2-aws/commit/5c84161f38a2d5bc7fd687b651046e16f428bf84 already in  https://github.com/laserdisc-io/fs2-aws/commits/series/3.x.x branch.

The problem happens when there are only 2 cores and global ExecutionContext is bounded to two threads.

Kinesis tests are now having a deadlock with the only two threads that are available in that case:
* one that blocks on `unsafeRunSync()` https://github.com/laserdisc-io/fs2-aws/blob/series/3.x.x/fs2-aws/src/main/scala/fs2/aws/kinesis/Kinesis.scala#L86 when reading kinesis stream
* another one that blocks on this `await()` https://github.com/laserdisc-io/fs2-aws/blob/series/3.x.x/fs2-aws/src/test/scala/fs2/aws/kinesis/NewKinesisConsumerSpec.scala#L362

I think that would require either:
* Change the implementation to don't block on `Kinesis.scala`. What was the problem with previous `runAsync` that was returning a `SyncIO` and not blocking on `unsafeRunSync()`? I think blocking a global thread waiting for https://github.com/laserdisc-io/fs2-aws/blob/series/3.x.x/fs2-aws/src/main/scala/fs2/aws/kinesis/Kinesis.scala#L86) is not the best approach
* Provide an ExecutionContext with at least 3 threads
* Run tests in github actions giving using `-Dscala.concurrent.context.numThreads=3` java property

Reply:
thanks for the investigation, that makes sense, i will fix this

Reply:
there is a reason we put this change. 
The previous semantic was fire-and-forget, meaning we submit the task to enqueue new message to the queue and do not block.
In case of slow consumer - fast producer we will keep submitting new tasks until OOM kick in.

Reply:
`ConcurrentEffect[F].toIO(queue.enqueue1(records)).unsafeRunSync()` 
this code is always executed in AWS SDK thread pool, not the one that CATs uses

Reply:
i believe we need to fix tests to use another cached context for AWS mocks